CHK-599: Fix rendering of markers not updating on address change

lucasecdb opened the pull request:
#### What is the purpose of this pull request?

This pull-request updates the rendering logic for the Map component to update 
the marker icon whenever the user changes their address. This actually is a 
much bigger refactor than I initially planned, but I've done it this way so the 
component is more maintainable across changes.

Basically, I re-implemented how the Map component works. I took much 
inspiration by the [react-google-maps](https://github.com/tomchentw/react-google-maps) library and 
its API. I created the `GoogleMap` and `GoogleMarker` components that were 
necessary for our use case, and moved all the rendering logic for markers into 
the `GoogleMarker` component, so we can render it declaratively instead of 
imperatively using refs.

So now, instead of needing to manually create the `google.maps.Marker` object 
and attach it to the map by calling `marker.setMap(this.map)`, we can simply 
render the `<GoogleMarker />` component as a children of `<GoogleMap />` and it 
will work ootb.

```tsx
<GoogleMap>
  <GoogleMarker location={[lng, lat]} url={markerIcon} />
</GoogleMap>
```

With this, I deleted most of the code inside the `Map` component which was 
related to this imperative rendering, and replaced it with the components, 
which will now be handled by React itself.  I believe it is now a much simpler 
component to understand and to change (going from ~800 LOC down to ~300).

But this change isn't without is compromises, now that I basically 
re-implemented it, there are some flows I might not be aware of that was 
handled with the old code that could not be handled with the new one.

~I still pretend to run the e2e tests with my changes, but haven't done that 
yet.~

Update: I've executed the e2e tests and they all passed, with the exception of two
tests that are always failing.

#### What problem is this solving?

Fixes the updating of marker icons when the address is changed. Before this, 
whenever you updated your address, the best pickup points of the previous 
location would still appear highlighted and the new ones wouldn't.

#### How should this be manually tested?

[Workspace](https://lucas--paguemenos.myvtex.com/checkout/cart/add/?sku=19449&qty=1&seller=1&sc=1)

You can use the postal code `60425-812`.

To see the issue resolved, simply change to a new address in Fortaleza and all 
the markers on the map should be updated accordingly. You can use for example 
the address `av treze de maio 1383`.

#### Screenshots or example usage

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

wdsrocha commented:
Lint is failing just because it is running on all files, right?

lucasecdb commented:
> Lint is failing just because it is running on all files, right?

yes, I fixed it in another pull-request #176

wdsrocha commented on react/components/GoogleMap.tsx:
```diff
@@ -0,0 +1,373 @@
+      const [referenceCenterLng, referenceCenterLat] =
+        referenceCenterRef.current
+
+      const differenceLat = (referenceCenterLat - currentCenter.lat()) * 1000
+      const differenceLng = (referenceCenterLng - currentCenter.lng()) * 1000
+
+      const DISTANCE = 50
+
+      const outerLat = differenceLat > DISTANCE || differenceLat < -DISTANCE
+      const outerLng = differenceLng > DISTANCE || differenceLng < -DISTANCE
+
+      setShouldSearchArea(outerLat || outerLng)
```
Have you considered using `Math.abs`? I believe it makes the code easier to understand

Besides that, I would put the constant on the top like the other contestants and would also remove the `outer` constants and make the comparison inside the `setShouldSearchArea`, but that's more of a personal preference. Wdyt?

```suggestion
      const differenceLat = Math.abs(referenceCenterLat - currentCenter.lat()) * 1000
      const differenceLng = Math.abs(referenceCenterLng - currentCenter.lng()) * 1000

      setShouldSearchArea(differenceLat > DISTANCE || differenceLng > DISTANCE)
```

lucasecdb replied:
makes sense to me! I haven't put much thought in this code because I didn't wanted to break something that was already working 😅

wdsrocha replied:
ahh no worries. I thought that this particular file might have been created from scratch

wdsrocha commented on react/components/GoogleMap.tsx:
```diff
@@ -0,0 +1,376 @@
+ */
+const PAN_LEFT_LNG = -30
+
+const MAP_STYLES: CSSProperties = {
+  height: '100%',
+  width: '100%',
+  position: 'absolute',
+  top: 0,
+  zIndex: 0,
+}
```
Nice way of using inline styles and arguably have a better DX than using standard CSS files (for this case, at least) ✨

wdsrocha reviewed (dismissed):
LGTM! Amazing job with the refactor 🥺

wdsrocha commented:
@lucasecdb what changed? Some main files changed since I last reviewed, but I'm not sure what 👀

lucasecdb commented on react/components/GoogleMarker.tsx:
```diff
@@ -0,0 +1,165 @@
+  }, [marker, icon, zIndex])
+
+  useMarkerEvent(marker, 'click', () => {
+    pickupPointSelectionEvent({ selectionMethod: SELECTION_METHOD_MAP })
```
this log was moved here from #177 (which placed this in `Map.jsx`)

lucasecdb commented on react/components/Map.jsx:
```diff
@@ -0,0 +1,314 @@
+              draggable={false}
+              zIndex={index < BEST_PICKUPS_AMOUNT ? FORTH_ZINDEX : THIRD_ZINDEX}
+              onClick={() => {
+                pickupPointSelectionEvent({
+                  selectionMethod: SELECTION_METHOD_MAP,
+                })
```
rebased the log added on #177 and moved it here

lucasecdb commented:
> @lucasecdb what changed? Some main files changed since I last reviewed, but I'm not sure which LOCs 👀

@wdsrocha I rebased the changes of #177 basically, I moved the logs from the markers to the `onClick` handlers of the `GoogleMarker` component, but I think the rest is all the same

ArthurSampaio commented on react/components/GoogleMap.tsx:
```diff
@@ -0,0 +1,376 @@
+        return
+      }
+
+      // TODO: can we move this `querySelector` call to use React refs instead?
```
It makes sense to me. What do you think about opening a new task and attaching it to our backlog?

lucasecdb replied:
I worry that it won't ever be prioritized for us to fix, so it would be just a "dead" task :-/

ArthurSampaio replied:
But it will be better tracked on the board than on this TODO. If you don't wanna open a new task, you can just delete this comment.

ArthurSampaio replied:
Also, this is a good example of a `good first bug` task to put to someone do in the onboarding process.

lucasecdb replied:
you convinced me 😄 created here: https://vtex-dev.atlassian.net/browse/CHK-1098

ArthurSampaio reviewed (approved):
lack changelog :)

lbebber commented on react/components/Map.jsx:
```diff
@@ -0,0 +1,317 @@
+const MARKER_HEIGHT = 31
+
+const THIRD_ZINDEX = 3
+const FORTH_ZINDEX = 4
```
```suggestion
const FOURTH_ZINDEX = 4
```

lbebber commented on react/components/Map.jsx:
```diff
@@ -0,0 +1,317 @@
+              key={pickupId}
+              position={location}
+              draggable={false}
+              zIndex={index < BEST_PICKUPS_AMOUNT ? FORTH_ZINDEX : THIRD_ZINDEX}
```
```suggestion
              zIndex={index < BEST_PICKUPS_AMOUNT ? FOURTH_ZINDEX : THIRD_ZINDEX}
```

lbebber reviewed (approved):
Some nitpicks, but overall LGTM, although admittedly this PR is a bit too large to be sure that everything is covered.

Great job on the refactor nonetheless! This seems like definitely an improvement in architecture

lucasecdb commented on react/PickupPointsModal.js:
```diff
@@ -265,25 +275,23 @@ class PickupPointsModal extends Component {
-                  mapStatus={mapStatus}
-                  shouldShow={shouldShowMap}
-                />
+                <div id="controls-wrapper" />
```
yep, I use it as a target for a react portal, because I needed to move the `ZoomControls` component to be a child of `Map` but the actual content of the zoom component must be placed in this div position

not the best solution, not proud of this 😂

lbebber replied:
ah I see, maybe add a lil comment right above it

lucasecdb replied:
done :-)

lucasecdb commented on react/components/GoogleMap.tsx:
```diff
@@ -0,0 +1,376 @@
+        return
+      }
+
+      if (center.length !== 2) {
```
I think I added this check while trying to debug an infinite loop to safe-guard and guarantee that this array was an actual geocoordinate

I'll add the utility function :-)
